Overload return type of `recv`

Currently `recv` and `recv_streaming` are typed as always returning `Data`, even if `decode` is set. This means that to satisfy type checkers, we need to use `isinstance`, `cast`, etc.

```
async for chunk in ws.recv_streaming(decode=False):
    assert isinstance(chunk, bytes)
    ...
```

I propose using `typing.overload` to narrow the return type when `decode` is not `None`:

```
@overload
async def recv(self, decode: Literal[True]) -> str:
    ...
@overload
async def recv(self, decode: Literal[True]) -> bytes:
    ...
@overload
async def recv(self, decode: None) -> Data:
    ...
async def recv(self, decode: bool | None = None) -> Data:
    """(actual implementation)"""
```

Alternatively, new `recv_str`/`recv_bytes` methods can automatically raise an Exception if the received data is not of the expected type.

I tried it but mypy doesn't survive the combination of async iteration and type overloads, as you can see in the CI run:

```
mypy --strict src
src/websockets/asyncio/messages.py:199: error: Overloaded function implementation cannot produce return type of signature 1  [misc]
src/websockets/asyncio/messages.py:199: error: Overloaded function implementation cannot produce return type of signature 2  [misc]
src/websockets/asyncio/messages.py:199: error: Overloaded function implementation cannot produce return type of signature 3  [misc]
src/websockets/asyncio/connection.py:339: error: Overloaded function implementation cannot produce return type of signature 1  [misc]
src/websockets/asyncio/connection.py:339: error: Overloaded function implementation cannot produce return type of signature 2  [misc]
src/websockets/asyncio/connection.py:339: error: Overloaded function implementation cannot produce return type of signature 3  [misc]
src/websockets/asyncio/connection.py:384: error: "Coroutine[Any, Any, AsyncIterator[str | bytes]]" has no attribute "__aiter__" (not async iterable)  [attr-defined]
src/websockets/asyncio/connection.py:384: note: Maybe you forgot to use "await"?
Found 7 errors in 2 files (checked 45 source files)
```

This looks like https://github.com/python/mypy/issues/10301.

When the default values aren't included in the overloads — i.e. when I follow your original suggestion — I get this error:

```
src/websockets/asyncio/connection.py:242: error: All overload variants of "recv" of "Connection" require at least one argument  [call-overload]
src/websockets/asyncio/connection.py:242: note: Possible overload variants:
src/websockets/asyncio/connection.py:242: note:     def recv(self, decode: Literal[True]) -> Coroutine[Any, Any, str]
src/websockets/asyncio/connection.py:242: note:     def recv(self, decode: Literal[False]) -> Coroutine[Any, Any, bytes]
src/websockets/asyncio/connection.py:242: note:     def recv(self, decode: None) -> Coroutine[Any, Any, str | bytes]
src/websockets/asyncio/connection.py:303: error: Argument 1 to "get" of "Assembler" has incompatible type "bool | None"; expected "None"  [arg-type]
src/websockets/asyncio/connection.py:378: error: Argument 1 to "get_iter" of "Assembler" has incompatible type "bool | None"; expected "None"  [arg-type]
src/websockets/sync/connection.py:240: error: All overload variants of "recv" of "Connection" require at least one argument  [call-overload]
src/websockets/sync/connection.py:240: note: Possible overload variants:
src/websockets/sync/connection.py:240: note:     def recv(self, timeout: float | None, decode: Literal[True]) -> str
src/websockets/sync/connection.py:240: note:     def recv(self, timeout: float | None, decode: Literal[False]) -> bytes
src/websockets/sync/connection.py:240: note:     def recv(self, timeout: float | None, decode: None) -> str | bytes
src/websockets/sync/connection.py:302: error: Argument 2 to "get" of "Assembler" has incompatible type "bool | None"; expected "None"  [arg-type]
src/websockets/sync/connection.py:372: error: Argument 1 to "get_iter" of "Assembler" has incompatible type "bool | None"; expected "None"  [arg-type]
```

I'm still running in trouble with this one:

```python
def get(self, timeout: float | None = None, decode: bool | None = None) -> Data:
```

- If I remove all the default values, mypy is unhappy.
- If I leave the default values (` = None`) in the overload, mypy is happy but you tell me that Pyright isn't.
- If I remove them only for the `Literal[True]` and `Literal[False]` case, then I get this error:

```
mypy --strict src
src/websockets/sync/connection.py:245: error: parameter without a default follows parameter with a default  [syntax]
```

--- a/docs/project/changelog.rst
+++ b/docs/project/changelog.rst
@@ -38,6 +38,12 @@ New features
 * Added :doc:`keepalive and latency measurement <../topics/keepalive>` to the
   :mod:`threading` implementation.
 
+Improvements
+............
+
+* Added type overloads for the ``decode`` argument of
+  :meth:`~asyncio.connection.Connection.recv`. This may simplify static typing.
+
 .. _14.2:
 
 14.2

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -67,7 +67,7 @@ exclude_lines = [
     "except ImportError:",
     "if self.debug:",
     "if sys.platform != \"win32\":",
-    "if typing.TYPE_CHECKING:",
+    "if TYPE_CHECKING:",
     "raise AssertionError",
     "self.fail\\(\".*\"\\)",
     "@unittest.skip",

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -70,6 +70,7 @@ exclude_lines = [
     "if TYPE_CHECKING:",
     "raise AssertionError",
     "self.fail\\(\".*\"\\)",
+    "@overload",
     "@unittest.skip",
 ]
 partial_branches = [

--- a/src/websockets/__init__.py
+++ b/src/websockets/__init__.py
@@ -1,6 +1,7 @@
 from __future__ import annotations
 
-import typing
+# Importing the typing module would conflict with websockets.typing.
+from typing import TYPE_CHECKING
 
 from .imports import lazy_import
 from .version import version as __version__  # noqa: F401
@@ -72,7 +73,7 @@ __all__ = [
 ]
 
 # When type checking, import non-deprecated aliases eagerly. Else, import on demand.
-if typing.TYPE_CHECKING:
+if TYPE_CHECKING:
     from .asyncio.client import ClientConnection, connect, unix_connect
     from .asyncio.server import (
         Server,

--- a/src/websockets/asyncio/connection.py
+++ b/src/websockets/asyncio/connection.py
@@ -11,7 +11,7 @@ import traceback
 import uuid
 from collections.abc import AsyncIterable, AsyncIterator, Awaitable, Iterable, Mapping
 from types import TracebackType
-from typing import Any, cast
+from typing import Any, Literal, cast, overload
 
 from ..exceptions import (
     ConcurrencyError,
@@ -243,6 +243,15 @@ class Connection(asyncio.Protocol):
         except ConnectionClosedOK:
             return
 
+    @overload
+    async def recv(self, decode: Literal[True]) -> str: ...
+
+    @overload
+    async def recv(self, decode: Literal[False]) -> bytes: ...
+
+    @overload
+    async def recv(self, decode: bool | None = None) -> Data: ...
+
     async def recv(self, decode: bool | None = None) -> Data:
         """
         Receive the next message.
@@ -311,6 +320,15 @@ class Connection(asyncio.Protocol):
         # Wait for the protocol state to be CLOSED before accessing close_exc.
         await asyncio.shield(self.connection_lost_waiter)
         raise self.protocol.close_exc from self.recv_exc
+
+    @overload
+    def recv_streaming(self, decode: Literal[True]) -> AsyncIterator[str]: ...
+
+    @overload
+    def recv_streaming(self, decode: Literal[False]) -> AsyncIterator[bytes]: ...
+
+    @overload
+    def recv_streaming(self, decode: bool | None = None) -> AsyncIterator[Data]: ...
 
     async def recv_streaming(self, decode: bool | None = None) -> AsyncIterator[Data]:
         """

--- a/src/websockets/asyncio/messages.py
+++ b/src/websockets/asyncio/messages.py
@@ -4,7 +4,7 @@ import asyncio
 import codecs
 import collections
 from collections.abc import AsyncIterator, Iterable
-from typing import Any, Callable, Generic, TypeVar
+from typing import Any, Callable, Generic, Literal, TypeVar, overload
 
 from ..exceptions import ConcurrencyError
 from ..frames import OP_BINARY, OP_CONT, OP_TEXT, Frame
@@ -116,6 +116,15 @@ class Assembler:
         # This flag marks the end of the connection.
         self.closed = False
 
+    @overload
+    async def get(self, decode: Literal[True]) -> str: ...
+
+    @overload
+    async def get(self, decode: Literal[False]) -> bytes: ...
+
+    @overload
+    async def get(self, decode: bool | None = None) -> Data: ...
+
     async def get(self, decode: bool | None = None) -> Data:
         """
         Read the next message.
@@ -175,6 +184,15 @@ class Assembler:
             return data.decode()
         else:
             return data
+
+    @overload
+    def get_iter(self, decode: Literal[True]) -> AsyncIterator[str]: ...
+
+    @overload
+    def get_iter(self, decode: Literal[False]) -> AsyncIterator[bytes]: ...
+
+    @overload
+    def get_iter(self, decode: bool | None = None) -> AsyncIterator[Data]: ...
 
     async def get_iter(self, decode: bool | None = None) -> AsyncIterator[Data]:
         """

--- a/src/websockets/sync/connection.py
+++ b/src/websockets/sync/connection.py
@@ -10,7 +10,7 @@ import time
 import uuid
 from collections.abc import Iterable, Iterator, Mapping
 from types import TracebackType
-from typing import Any
+from typing import Any, Literal, overload
 
 from ..exceptions import (
     ConcurrencyError,
@@ -241,6 +241,28 @@ class Connection:
         except ConnectionClosedOK:
             return
 
+    # This overload structure is required to avoid the error:
+    # "parameter without a default follows parameter with a default"
+
+    @overload
+    def recv(self, timeout: float | None, decode: Literal[True]) -> str: ...
+
+    @overload
+    def recv(self, timeout: float | None, decode: Literal[False]) -> bytes: ...
+
+    @overload
+    def recv(self, timeout: float | None = None, *, decode: Literal[True]) -> str: ...
+
+    @overload
+    def recv(
+        self, timeout: float | None = None, *, decode: Literal[False]
+    ) -> bytes: ...
+
+    @overload
+    def recv(
+        self, timeout: float | None = None, decode: bool | None = None
+    ) -> Data: ...
+
     def recv(self, timeout: float | None = None, decode: bool | None = None) -> Data:
         """
         Receive the next message.
@@ -310,6 +332,15 @@ class Connection:
         # Wait for the protocol state to be CLOSED before accessing close_exc.
         self.recv_events_thread.join()
         raise self.protocol.close_exc from self.recv_exc
+
+    @overload
+    def recv_streaming(self, decode: Literal[True]) -> Iterator[str]: ...
+
+    @overload
+    def recv_streaming(self, decode: Literal[False]) -> Iterator[bytes]: ...
+
+    @overload
+    def recv_streaming(self, decode: bool | None = None) -> Iterator[Data]: ...
 
     def recv_streaming(self, decode: bool | None = None) -> Iterator[Data]:
         """

--- a/src/websockets/sync/messages.py
+++ b/src/websockets/sync/messages.py
@@ -3,7 +3,7 @@ from __future__ import annotations
 import codecs
 import queue
 import threading
-from typing import Any, Callable, Iterable, Iterator
+from typing import Any, Callable, Iterable, Iterator, Literal, overload
 
 from ..exceptions import ConcurrencyError
 from ..frames import OP_BINARY, OP_CONT, OP_TEXT, Frame
@@ -110,6 +110,24 @@ class Assembler:
             for frame in queued:  # pragma: no cover
                 self.frames.put(frame)
 
+    # This overload structure is required to avoid the error:
+    # "parameter without a default follows parameter with a default"
+
+    @overload
+    def get(self, timeout: float | None, decode: Literal[True]) -> str: ...
+
+    @overload
+    def get(self, timeout: float | None, decode: Literal[False]) -> bytes: ...
+
+    @overload
+    def get(self, timeout: float | None = None, *, decode: Literal[True]) -> str: ...
+
+    @overload
+    def get(self, timeout: float | None = None, *, decode: Literal[False]) -> bytes: ...
+
+    @overload
+    def get(self, timeout: float | None = None, decode: bool | None = None) -> Data: ...
+
     def get(self, timeout: float | None = None, decode: bool | None = None) -> Data:
         """
         Read the next message.
@@ -180,6 +198,15 @@ class Assembler:
             return data.decode()
         else:
             return data
+
+    @overload
+    def get_iter(self, decode: Literal[True]) -> Iterator[str]: ...
+
+    @overload
+    def get_iter(self, decode: Literal[False]) -> Iterator[bytes]: ...
+
+    @overload
+    def get_iter(self, decode: bool | None = None) -> Iterator[Data]: ...
 
     def get_iter(self, decode: bool | None = None) -> Iterator[Data]:
         """

--- a/src/websockets/typing.py
+++ b/src/websockets/typing.py
@@ -2,8 +2,7 @@ from __future__ import annotations
 
 import http
 import logging
-import typing
-from typing import Any, NewType, Optional, Union
+from typing import TYPE_CHECKING, Any, NewType, Optional, Union
 
 
 __all__ = [
@@ -31,7 +30,7 @@ Data = Union[str, bytes]
 
 
 # Change to logging.Logger | ... when dropping Python < 3.10.
-if typing.TYPE_CHECKING:
+if TYPE_CHECKING:
     LoggerLike = Union[logging.Logger, logging.LoggerAdapter[Any]]
     """Types accepted where a :class:`~logging.Logger` is expected."""
 else:  # remove this branch when dropping support for Python < 3.11

